### ISSUE TITLE
`dataclass`es that inherit from `Protocol` subclasses have wrong `__init__`

### ISSUE THREAD

**Issue by julianfortune** [python/typing#883]
I believe [`bpo-44806: Fix __init__ in subclasses of protocols`](https://github.com/python/cpython/pull/27545) has caused a regression when using a Dataclass.

In Python `3.9.7`, a `dataclass` that inherits from a subclass of `typing.Protocol` (i.e., a user-defined protocol), does not have the correct `__init__`.

### Demonstration

```python
from dataclasses import dataclass
from typing import Protocol

class P(Protocol):
    pass

@dataclass
class B(P):
    value: str

print(B("test"))
```
In `3.9.7`:
```shell
Traceback (most recent call last):
  File "test.py", line 11, in <module>
    print(B("test"))
TypeError: B() takes no arguments
```
In `3.9.6`:
```shell
B(value='test')
```

### Affected Projects

- [dbt](https://github.com/dbt-labs/dbt/issues/3843)

**Comment by srittau** [comment]
Please report this at bugs.python.org.

**Comment by julianfortune** [comment]
Thanks for the fast response.

[Python Issues](https://bugs.python.org/issue45081)

**Comment by AthenaCaesura** [comment]
This bug was solved in python 3.10, however the issue still remains for 3.9.7.

Is there any workaround that one could use instead of upgrading? Since 3.9.7 is the last version of 3.9 saying we don't support 3.9.7 is more or less the same as saying we don't support 3.9.

I would like to keep the frozen option if at all possible.

**Comment by hauntsaninja** [comment]
Python 3.9.10 is the latest release of 3.9 and presumably contains https://github.com/python/cpython/pull/28132
